[RM Appendix A] Add VNF Design requirements (#1228)
cntt-n/CNTT: doc/ref_model/chapters/appendix-a.md

[<< Back](../../ref_model)

> # Appendix A - VNF Design Guidelines

Reviewer: ```suggestion
# Appendix A - Guidelines For VNF Vendors	
```
The title was changed from "VNF Design Guidelines" to "Guidelines for VNF Vendors" back in 2019.  This title seems more appropriate, as the guidelines here expand a bit more beyond just design guidelines.  For instance, the VNF evolution phases provides some background on CNTT design principles.

Reviewer 2: Agree. The very original idea was that all guidance to VNF vendors goes into that one place, and text on impact between VNF design and platform is great fit there.

Author: @kedmison References to the Appendix in RM as well as the Title of the App use "VNF Design Guidelines"

Changing Appendix Title (within doc and TOC) to "Guidelines for VNF Vendors" but please ensure changes are made in all references to this Appendix.

Reviewer: Thanks for flagging this.  PR #1293 is created.

End of policies.

<a name="A.6"></a>

> ## A.6 VNF Design and Implementation Requirements

Reviewer: ```suggestion
## A.6 VNF Design and Implementation Guidelines
```

Author: @kedmison These are Requirements. I think we should include the ONAP "[VNF or PNF Requirements Documentation](https://onap.readthedocs.io/en/latest/submodules/vnfrqts/requirements.git/docs/index.html)"

Reviewer: I believe the CNTT should focus on standardizing the VNF<->NFVI interface.  I believe it's OK to make observations about VNF design that best takes advantage of this VNF-NFVI interface, but requirements about the process for building the VNF, or the contents of it, are probably best left to other forum like ONAP, etc.  

The other reason is that if it is made mandatory here how is it going to be testable?  Is the RC going to mandate these requirements, and then be required to execute tests for them? If not, then the requirement is not really a 'requirement' and may serve to weaken the RM document instead.

> This section specifies a set of VNF Design and Implementation Requirements. 

Reviewer: ```suggestion
This section specifies a set of VNF Design and Implementation Guidelines.   While not mandatory, VNF vendors are strongly encouraged to comply with these guidelines.
```

For the most part, the CNTT does not have a mechanism for enforcing them as 'must' requirements.  Because they are mostly un-testable at RC time, it is likely better to specify them as suggested/strongly encouraged tasks.

Author: @kedmison I think we need both requirements and guidelines.

> | Ref | Requirmenet | Unit | Definition

Reviewer: ```suggestion
| Ref | Requirement | Unit | Definition
```

Author: @kedmison Thanks -- Corrected

|----|-----|-----|----|

> | vnf.des.001 | The VNF Developer **must** ensure that their software and the resultant VNF image does not contain malicious code (e.g., malware, logic bombs, etc.)  |  | |

Reviewer: ```suggestion
| vnf.des.001 | The VNF Developer **should** ensure that their software and the resultant VNF image does not contain malicious code (e.g., malware, logic bombs, etc.)  |  | |
```

> | vnf.des.002 | The VNF Developer **must** ensure that their software and the resultant VNF image does not contain code such as daemons that exposes them to risk |  | |

Reviewer: ```suggestion
| vnf.des.002 | The VNF Developer **should** ensure that their software and the resultant VNF image does not contain code such as daemons that exposes them to risk |  | |
```

Author: @kedmison Why should this not be required? Operators could ask for Vendors to certify to this.

Reviewer: Indeed operators can do this.  But if it is specified here as a must, then it has to be flowed through to RI/RC as a requirement to test.

Perhaps this is best left as a 'should', and operators can make some of these mandatory during their procurement cycles if so desired.

> | vnf.des.003 | The VNF Developer **must** ensure that their software and resultant VNF image are only created with content and files from trusted sources |  | |

Reviewer: ```suggestion
| vnf.des.003 | The VNF Developer **should** ensure that their software and resultant VNF image are only created with content and files from trusted sources |  | |
```

Author: @kedmison Why should this not be required? Operators could ask for Vendors to certify to this.

Reviewer: How could RC ensure that a VNF is conformant to this?  At validation time, the RC would not know what the source of the content is... only the VNF vendor knows.  

Operators can make some of these mandatory during their procurement cycles if so desired.

> | vnf.des.004 | The VNF Developer **must** ensure that their software and resultant VNF image are only packaged with files that have been found free of malware and vulnerabilities |  | |

Reviewer: ```suggestion
| vnf.des.004 | The VNF Developer **should** ensure that their software and resultant VNF image are only packaged with files that have been found free of malware and vulnerabilities |  | |
```

> | vnf.des.005 | The VNF Developer **must** ensure that their software and resultant VNF image do not contain clear text secrets |  | |

Reviewer: ```suggestion
| vnf.des.005 | The VNF Developer **should** ensure that their software and resultant VNF image do not contain clear text secrets |  | |
```

Author: @kedmison Why should this not be required? Operators could ask for Vendors to certify to this.

Reviewer: Indeed it can be.  But how would it be made a testable requirement at the RC?
Operators can make some of these mandatory during their procurement cycles if so desired.

| vnf.des.006 | The VNF Developer **should** follow the guidance in the CSA Security Guidance for Critical Areas of Focus in Cloud Computing (latest version) [https://cloudsecurityalliance.org/](https://cloudsecurityalliance.org/) |  | Cloud Security Alliance - [https://cloudsecurityalliance.org/](https://cloudsecurityalliance.org/) |
| vnf.des.007 | The VNF Developer **should** follow the guidance in the OWASP Cheat Sheet Series (OCSS) https://github.com/OWASP/CheatSheetSeries |  | Open Web Application Security Project [https://www.owasp.org](https://www.owasp.org) |
| vnf.des.008 | The VNF Developer **should** ensure that their code is not vulnerable to the OWASP Top Ten Security Risks [https://owasp.org/www-project-top-ten/](https://owasp.org/www-project-top-ten/) |  |

<a name="A.7"></a>

> ## A.6 Miscellaneous

Reviewer: ```suggestion
## A.7 Miscellaneous
```

> ### A.6.1 VNF Network Monitoring Capabilities - UseCase.

Reviewer: ```suggestion
### A.7.1 VNF Network Monitoring Capabilities - UseCase.
```
